src/engine/enginebuffer.cpp

```diff
@@ -1191,20 +1188,19 @@ void EngineBuffer::processSlip(int iBufferSize) {
     if (enabled != m_bSlipEnabledProcessing) {
         m_bSlipEnabledProcessing = enabled;
         if (enabled) {
-            m_dSlipPosition = m_filepos_play;
+            m_slipPosition = m_playPosition;
             m_dSlipRate = m_rate_old;
         } else {
             // TODO(owen) assuming that looping will get canceled properly
-            double newPlayFrame = m_dSlipPosition / kSamplesPerFrame;
-            double roundedSlip = round(newPlayFrame) * kSamplesPerFrame;
-            slotControlSeekExact(roundedSlip);
-            m_dSlipPosition = 0;
+            seekExact(m_slipPosition.toNearestFrameBoundary());
+            m_slipPosition = mixxx::audio::kStartFramePos;
         }
     }
 
     // Increment slip position even if it was just toggled -- this ensures the position is correct.
     if (enabled) {
-        m_dSlipPosition += static_cast<double>(iBufferSize) * m_dSlipRate;
+        m_slipPosition += static_cast<double>(iBufferSize) * m_dSlipRate /
```

Comment:
Use `FramePos::fromEngineSamplePos()`

Reply:
Addition of FramePos is undefined, because both share the same origin, so it doesn't make sense semantically.

Reply:
We could hack around it by doing the addition with (position - startpos) but that is less clear IMHO.

Reply:
`m_slipPosition = FramePos::fromEngineSamplePos(static_cast<double>(iBufferSize) * m_dSlipRate);`?

Reply:
Didn't notice the +=. Nevertheless we can use it.

Reply:
I agree that it's not nice to hardcode the channel number here, but it seems kind of hacky to convert it to a framepos first and then getting it raw value.

I also noticed that the buffer size is guaranteed to be a multiple of 2, so we can use integer division here. Please have a look at https://github.com/mixxxdj/mixxx/pull/4217/commits/bc616ccb8f987cedd7a2283a16ea658489cc3d7a and let me know if this is okay. In the long term, we should check if we even need the buffer size or can use the frames count directly.